File: app/src/main/kotlin/tech/mmarca/openvitals/ui/components/PersonalBaselineStat.kt
package tech.mmarca.openvitals.ui.components

import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.outlined.TrendingDown
import androidx.compose.material.icons.automirrored.outlined.TrendingFlat
import androidx.compose.material.icons.automirrored.outlined.TrendingUp
import androidx.compose.material.icons.outlined.CalendarMonth
import androidx.compose.material.icons.outlined.Star
import androidx.compose.runtime.Composable
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.stringResource
import tech.mmarca.openvitals.R
import tech.mmarca.openvitals.core.insights.BaselineStatus
import tech.mmarca.openvitals.core.insights.PersonalBaselineInsight
import tech.mmarca.openvitals.core.presentation.DisplayValue
import tech.mmarca.openvitals.core.presentation.UnitFormatter
import kotlin.math.abs
import kotlin.math.roundToInt

@Composable
fun personalBaselineInsightStats(
    insight: PersonalBaselineInsight?,
    unitFormatter: UnitFormatter,
    valueFormatter: @Composable (Double) -> DisplayValue,
    accentColor: Color,
): List<InsightStat> {
    if (insight == null) return emptyList()

    val baselineStats = insight.summaries.map { summary ->
        val average = valueFormatter(summary.average)
        InsightStat(
            title = stringResource(
                when (summary.windowDays) {
                    30 -> R.string.stat_30_day_baseline
                    60 -> R.string.stat_60_day_baseline
                    90 -> R.string.stat_90_day_baseline
                    else -> R.string.stat_baseline
                }
            ),
            value = average.value,
            unit = average.unit,
            icon = Icons.Outlined.CalendarMonth,
            accentColor = accentColor,
        )
    }

    val deviation = insight.percentDeviation
        ?.let { percent ->
            DisplayValue(
                value = signedValue(
                    value = unitFormatter.count(abs(percent).roundToInt()),
                    status = insight.status,
                ),
                unit = stringResource(R.string.unit_percent_symbol),
            )
        }
        ?: valueFormatter(insight.absoluteDeviation).let { absolute ->
            DisplayValue(
                value = signedValue(absolute.value, insight.status),
                unit = absolute.unit,
            )
        }

    return baselineStats + listOf(
        InsightStat(
            title = stringResource(R.string.stat_usual_range),
            value = baselineStatusLabel(insight.status),
            unit = "",
            icon = when (insight.status) {
                BaselineStatus.USUAL -> Icons.AutoMirrored.Outlined.TrendingFlat
                BaselineStatus.ABOVE,
                BaselineStatus.UNUSUAL_HIGH -> Icons.AutoMirrored.Outlined.TrendingUp
                BaselineStatus.BELOW,
                BaselineStatus.UNUSUAL_LOW -> Icons.AutoMirrored.Outlined.TrendingDown
            },
            accentColor = accentColor,
        ),
        InsightStat(
            title = stringResource(R.string.stat_baseline_deviation),
            value = deviation.value,
            unit = deviation.unit,
            icon = Icons.Outlined.Star,
            accentColor = accentColor,
        ),
    )
}

@Composable
private fun baselineStatusLabel(status: BaselineStatus): String =
    stringResource(
        when (status) {
            BaselineStatus.USUAL -> R.string.baseline_status_usual
            BaselineStatus.ABOVE -> R.string.baseline_status_above
            BaselineStatus.BELOW -> R.string.baseline_status_below
            BaselineStatus.UNUSUAL_HIGH -> R.string.baseline_status_unusual_high
            BaselineStatus.UNUSUAL_LOW -> R.string.baseline_status_unusual_low
        }
    )

private fun signedValue(value: String, status: BaselineStatus): String =
    when (status) {
        BaselineStatus.ABOVE,
        BaselineStatus.UNUSUAL_HIGH -> "+$value"
        BaselineStatus.BELOW,
        BaselineStatus.UNUSUAL_LOW -> "-$value"
        BaselineStatus.USUAL -> value
    }

File: lib/ui/components/personal_baseline_stat.dart
import 'package:flutter/material.dart';

import '../../core/presentation/display_value.dart';
import '../../core/presentation/unit_formatter.dart';
import '../../domain/insights/personal_baseline.dart';
import '../../l10n/app_localizations.dart';
import 'insight_cards.dart';

/// Port of the Kotlin `PersonalBaselineStat.kt`.

String _baselineStatusLabel(BaselineStatus status, AppLocalizations l10n) =>
    switch (status) {
      BaselineStatus.usual => l10n.baselineStatusUsual,
      BaselineStatus.above => l10n.baselineStatusAbove,
      BaselineStatus.below => l10n.baselineStatusBelow,
      BaselineStatus.unusualHigh => l10n.baselineStatusUnusualHigh,
      BaselineStatus.unusualLow => l10n.baselineStatusUnusualLow,
    };

/// Kotlin `signedValue(value, status)`.
String _signedValue(String value, BaselineStatus status) => switch (status) {
      BaselineStatus.above || BaselineStatus.unusualHigh => '+$value',
      BaselineStatus.below || BaselineStatus.unusualLow => '-$value',
      BaselineStatus.usual => value,
    };

IconData _statusIcon(BaselineStatus status) => switch (status) {
      BaselineStatus.usual => Icons.trending_flat,
      BaselineStatus.above || BaselineStatus.unusualHigh => Icons.trending_up,
      BaselineStatus.below || BaselineStatus.unusualLow => Icons.trending_down,
    };

/// Kotlin `personalBaselineInsightStats`: one card per baseline window, then the
/// usual-range verdict and the deviation. Empty when there is no baseline yet.
List<InsightStat> personalBaselineInsightStats({
  required PersonalBaselineInsight? insight,
  required UnitFormatter unitFormatter,
  required DisplayValue Function(double value) valueFormatter,
  required Color accentColor,
  required AppLocalizations l10n,
}) {
  if (insight == null) return const [];

  final stats = <InsightStat>[
    for (final summary in insight.summaries)
      InsightStat(
        title: switch (summary.windowDays) {
          30 => l10n.stat30DayBaseline,
          60 => l10n.stat60DayBaseline,
          90 => l10n.stat90DayBaseline,
          _ => l10n.statBaseline,
        },
        value: valueFormatter(summary.average).value,
        unit: valueFormatter(summary.average).unit,
        icon: Icons.calendar_month_outlined,
        accentColor: accentColor,
      ),
  ];

  final percent = insight.percentDeviation;
  final deviation = percent != null
      ? DisplayValue(
          _signedValue(unitFormatter.count(percent.abs().round()), insight.status),
          l10n.unitPercentSymbol,
        )
      : () {
          final absolute = valueFormatter(insight.absoluteDeviation);
          return DisplayValue(
            _signedValue(absolute.value, insight.status),
            absolute.unit,
          );
        }();

  return [
    ...stats,
    InsightStat(
      title: l10n.statUsualRange,
      value: _baselineStatusLabel(insight.status, l10n),
      unit: '',
      icon: _statusIcon(insight.status),
      accentColor: accentColor,
    ),
    InsightStat(
      title: l10n.statBaselineDeviation,
      value: deviation.value,
      unit: deviation.unit,
      icon: Icons.star_outline,
      accentColor: accentColor,
    ),
  ];
}
